### PR TITLE
[RISC-V] Opt `emitLoadImmediate`

### PR DESCRIPTION
**Pull request by clamp03**
I found unnecessary moves are generated on RISC-V.
example.
```
lui            a1, 516597
addiw          a1, a1, -93
slli           a1, a1, 7
mv             a1, a1
```

To remove such moves, the PR skips `slli` and `add`  in case of `low11` is zero in each iteration.
\*. I found the same optimization in `src/coreclr/vm/riscv64/stubs.cpp` and copied the codes.

Part of https://github.com/dotnet/runtime/issues/84834
cc @dotnet/samsung

### CODE REVIEW

**Comment by risc-vv** [comment]
<details><summary>e67f60f2 is being scheduled for building and testing</summary>

GIT: `e67f60f2406475a3e23af29676cdcff7a3e9f01b`
REPO: [dotnet/runtime](https://github.com/dotnet/runtime)
BRANCH: main</details>

File: src/coreclr/jit/emitriscv64.cpp [hunk 1]
```diff
@@ -1262,15 +1262,27 @@ void emitter::emitLoadImmediate(emitAttr size, regNumber reg, ssize_t imm)
...
     while (remainingShift > 0)
     {
         UINT32 shift = remainingShift >= 11 ? 11 : remainingShift % 11;
-        emitIns_R_R_I(INS_slli, size, reg, reg, shift);
+        UINT32 mask  = 0x7ff >> (11 - shift);
+        remainingShift -= shift;
+        ssize_t low11 = (imm >> remainingShift) & mask;
+        shiftAccumulator += shift;
+
+        if (low11)
```
Comment (tomeksowi) [inline src/coreclr/jit/emitriscv64.cpp:1275]: While you're here, shouldn't line 1261 `emitIns_R_R_I(INS_addiw, size, reg, reg, high31 & 0xFFF);` be under an analogous condition? I think immediates like 0x123000 also emit an unnecessary addiw.

Comment (clamp03) [reply src/coreclr/jit/emitriscv64.cpp:1277]: Thank you. Updated.

**Comment by risc-vv** [comment]
<details><summary>722baf4a is being scheduled for building and testing</summary>

GIT: `722baf4a4a0aa3d26b6d27b37ca796e41483cb79`
REPO: [dotnet/runtime](https://github.com/dotnet/runtime)
BRANCH: main</details>

File: src/coreclr/jit/emitriscv64.cpp [hunk 2]
```diff
@@ -1258,7 +1258,9 @@ void emitter::emitLoadImmediate(emitAttr size, regNumber reg, ssize_t imm)
     INT32 high19 = ((int32_t)(high31 + 0x800)) >> 12;
 
     emitIns_R_I(INS_lui, size, reg, high19);
-    emitIns_R_R_I(INS_addiw, size, reg, reg, high31 & 0xFFF);
+    if (high31 & 0xFFF) {
```
Comment (tomeksowi) [inline src/coreclr/jit/emitriscv64.cpp:1261]: Shouldn't the opening brace be on a new line?

File: src/coreclr/jit/emitriscv64.cpp [hunk 3]
```diff
@@ -1258,19 +1258,33 @@ void emitter::emitLoadImmediate(emitAttr size, regNumber reg, ssize_t imm)
...
+        emitIns_R_R_I(INS_addiw, size, reg, reg, high31 & 0xFFF);
+    }
 
     // And load remaining part part by batches of 11 bits size.
     INT32 remainingShift = msb - 30;
+
+    UINT32 shiftAccumulator = 0;
     while (remainingShift > 0)
     {
         UINT32 shift = remainingShift >= 11 ? 11 : remainingShift % 11;
```
Comment (tomeksowi) [inline src/coreclr/jit/emitriscv64.cpp:1271]: We could utilize all 12 bits by adding the sign bit like in line 1258. Up to you if this PR should include it.

**Review by tomeksowi — APPROVED** [review]
LGTM. No more suggestions to this opt.

File: src/coreclr/jit/emitriscv64.cpp (hunk 2 above)
Comment (clamp03) [reply src/coreclr/jit/emitriscv64.cpp]: Thank you. Fixed.

File: src/coreclr/jit/emitriscv64.cpp (hunk 3 above)
Comment (tomeksowi) [reply src/coreclr/jit/emitriscv64.cpp:1272]: Er, that would require a change to walk the immediate from least to most significant bits, I think we can do that later.

Comment (clamp03) [reply src/coreclr/jit/emitriscv64.cpp:1272]: I think 11 bits is enough because 12 bits cannot reduce instructions and it needs handling for sign bit.

**Comment by risc-vv** [comment]
<details><summary>51250ae7 is being scheduled for building and testing</summary>

GIT: `51250ae718b45231befa5c4011e741f73a1b5154`
REPO: [dotnet/runtime](https://github.com/dotnet/runtime)
BRANCH: main</details>

**Review by sirntar — APPROVED** [review]

**Comment by risc-vv** [comment]
<details><summary>d13b8b8b is being scheduled for building and testing</summary>

GIT: `d13b8b8b5c2ed298935771a4a60489a6d64eb934`
REPO: [dotnet/runtime](https://github.com/dotnet/runtime)
BRANCH: main</details>

**Comment by clamp03** [comment]
@jakobbotsch Could you take a look at this PR? Thank you!

**Comment by risc-vv** [comment]
<details><summary>d13b8b8b is being scheduled for building and testing</summary>

GIT: `d13b8b8b5c2ed298935771a4a60489a6d64eb934`
REPO: [dotnet/runtime](https://github.com/dotnet/runtime)
BRANCH: main</details>

**Review by jakobbotsch — APPROVED** [review]

**Pull request merged**